### ISSUE TITLE
How to replace the tile URL in a custom TileSource without disrupting the viewer?

### ISSUE THREAD

**Issue by bennettrogers** [openseadragon/openseadragon#1970]
I've been really enjoying working with OpenSeadragon!

I'm trying to figure out the best approach for a somewhat complicated feature. I'm using a custom TileSource that uses `getTileUrl` to fetch tiles from an S3 bucket. (I don't think it's relevant, but I'm also using range requests to get the tiles from a single image, so the tile URL is the same for every tile.) I'm trying to make it more secure by having the S3 signed URL expire after a fairly short time, which means I then need to update the TileSource to use this new URL. Somehow I need to seamlessly replace the URL without triggering a redraw in the viewer - it just needs to use the new URL whenever new tiles are fetched.

There are 2 approaches I'm considering:

1. Use `addTiledImage` to replace the TileSource, and somehow prevent the viewer from flickering/stuttering/freezing while the swap takes place. This seems a bit heavy handed, but probably fairly straightforward (although I haven't gotten it working yet).
2. Create a custom version of OSD that will retry a failed tile fetch (I've seen the retry logic in #1627, and would probably start there). I think I would also need to add some kind of `preRetry` function that gets called to replace the tile URL for the viewer. This seems more complicated, but possibly more robust.

In both cases, the problem boils down to finding a way to replace the URL for the TileSource without otherwise disrupting the viewer. Do you have any thoughts on the cleanest approach to accomplishing this?

Thank you for your work on this library!

**Comment by msalsbery** [comment]
This is a tough one currently...If I'm reading the code correctly, it looks like the URL and AJAX headers are tied up once an image's Tile objects are created and added to TiledImage.tilesMatrix. I don't see a public, documented way of manipulating TiledImage.tilesMatrix...

It's not uncommon for server APIs to update tokens periodically, so I think enhancing the AJAX tile loading support  with the ability to update tile URLs and/or AJAX headers on open images and their associated tiles is a reasonable enhancement. What do you think @iangilman ?

**Comment by iangilman** [comment]
@msalsbery agreed... having a smoother way to handle this scenario seems like a good enhancement!

**Comment by msalsbery** [comment]
@iangilman  Thanks! :)  

If I'm following correctly, the only tricky issue is when previously-loaded tiles need reloading - at that point they'd have the old URL. I think your idea would cover that.

And if I'm way off base, then never mind ;)

**Comment by bennettrogers** [comment]
@msalbery brings up a good point – using rotating tokens for authenticating image requests is probably a more common use case than what I'm trying to do, and it leads to this same issue. In either case, you'd need a way to update the URLs and/or headers for any future tile fetches, both for tiles that you've already cached and unseen tiles. 

@iangilman yes, it's possible to know ahead of time that the URL needs to be updated. You can get the expiration from the signed URL params, fetch the new URL before that time (assuming the tab is active?), and have `getTileURL` use the new value. When I tried to do this though, the viewer was still trying to fetch the old URL along with the new one. I haven't dug far enough into OSD yet to know if this is because it's trying to fetch a cached tile when you zoom in on an area that's already been fetched, or if I screwed up the URL update. It looks like the `cacheKey` for a tile is based on the URL (and the AJAX headers), so maybe that could be why.

I was also thinking more about approach #2 (retry on fail). My thought was to wait for the fetch to fail, then determine if it's because of an expired URL, and if so tell the event handler to trigger a URL update in the parent component and only then retry. I'd want to update the URL for every future tile from the first failure event handler, to prevent a barrage of failures/updates. Somehow this feels like the safer approach since it avoids trying to update the URL in a viewer that may not even be in an active tab (e.g. I don't know what happens to JS in background tabs in Chrome), yet it also feels like a worse UX since there would be a slight delay in tile loads every time the URL expires while waiting for the new one.

Maybe the best UX would be a combination of the 2 approaches - try to preemptively update the URL before it expires, but have a catch/update/retry mechanism as a fallback. It seems like the cache could be tricky in both approaches though.

I think I need to dig into the OSD source to understand exactly how the caching works and when the cached URLs are used. With the current cache implementation that uses the URL and headers as the `cacheKey`, it seems like it may be optimistic to try to update the URL and expect it to "just work". I wonder I can make a custom caching implementation that uses a combination of some unique image ID and tile level/x/y as the cacheKey.

I'm going to get more familiar with the OSD source so that I can speak more intelligently about this, but if you have other thoughts in the meantime I'd love to hear them!

**Comment by bennettrogers** [comment]
My current thought is that I can add an event handler for `update-tile`, in which I check the existing tile URL to see if it is expiring soon. If so, I can specify a new URL on the tile object that I think will update the object in the matrix - does that sound like it might work? I'm worried about any performance hit to calling a function on every single `update-tile` event, but maybe this is ok? What is the actual purpose of that event and is there a better one to use? Or perhaps I should add a `tile-loading` event that fires just before the tile is fetched.

One complication is still the `cacheKey`. Since it relies on the URL, setting a new URL on an existing tile will probably cause extra image fetches that don't need to happen, since the key won't match. I'm going to try adding a `getTileCacheKey` function to the `tileSource` and call that in `getTile` in the same place `getTileUrl` is called, and then pass this key to the `Tile` constructor.

**Comment by iangilman** [comment]
It looks like `update-tile` is called on every tile as it's considered for drawing, not just tiles that are about to be loaded, so that seems too broad. If you want to know when a tile is about to be loaded, adding a new event for that seems appropriate. 

It does seem like it might be worth exploring a new cache key system that's independent of URL. We just need to make sure it works for all scenarios. In your case you have the same image but the URL is changing. In some other cases (like a constantly-updated live feed) it may be the same URL but a new image. 

Anyway, sounds like you're on the right track. Good luck on your investigation! Please keep us posted :)

**Comment by bennettrogers** [comment]
I think it's working now.

The difficulty was that not only is the `tile.url` locked in when the tile is created, `tileSource.getTileUrl` is as well. So if my custom `getTileUrl` function relies on some updated (React) state to get the new URL for all future loaded tiles, it will still return the old URL based on the state from the point where `tileSource.getTileUrl` was created (during `viewer.open(tileSource)`).

The things I had to do were:
1) When my URL state changes, call `viewer.source.getTileUrl = getCurrentTileUrl`, where `getCurrentTileUrl` is the function I would have normally passed to `getTileUrl` in the `tileSource` options.
2) In my new `tile-loading` event handler, call `tile.url = tiledImage.source.getTileUrl(tile.level, tile.x, tile.y)`.
3) Add a custom `getTileCacheKey` function to the `tileSource`, call it inside `getTile`, and pass the result to the `Tile` constructor (using the default cacheKey implementation if the custom function is not defined). This lets me divorce the cacheKey from the (now dynamic) tile URL.

You can see the changes in [my fork of OSD](https://github.com/bennettrogers/openseadragon). I'd love to know if you think there's a better way, or if this is a terrible approach!

I still think I might need to add a `tile-load-failed` event handler, to handle any cases where the updated state didn't get noticed (maybe the tab was in the background when the URL expired, etc). If I end up doing that I'll update this issue.

**Comment by iangilman** [comment]
Seems like a reasonable approach! I don't think you necessarily need to replace the `getTileUrl`... can't you have it rely on the state of some variable? Something like this:

```
var extra = 'foo';

function getTileUrl(level, x, y) {
  return `${baseUrl}/${level}/${x}_${y}?extra=${extra}`;
}

extra = 'bar';
```

Regarding the cacheKey change, that seems like something that might be worth bringing into OSD, to support this sort of scenario elsewhere. Would you be up for making a pull request? I imagine there's some additional work to do to make a proper PR, but you seem to have made a good start so far!

**Comment by JohnReagan** [comment]
Hi, I am also hitting this issue. Same use case, would like to use presigned URLs from AWS to serve assets that expire after some time. What's the current thinking on the best way to handle this issue? 

Currently, I'm patching the library at runtime to allow the Tile.url field to accept either a string or function that returns a string. Then, my custom TileSource class's getTileUrl function returns a function so urls can be grabbed dynamically. Code snippet below: 

```typescript
Object.defineProperty(OpenSeadragon.Tile.prototype, 'url', {
  get() {
    const url: string | (() => string) = this._url;
    if (typeof url === 'function') {
      return url();
    }
    return url;
  },

  set(val) {
    this._url = val;
  },
});
```

My other approach involves iterating over the TiledImage.tilesMatrix structure to manually update all URLs whenever the old URL is invalidated. This feels a little brittle to race conditions. Is either approach recommended? Thanks for your support and the library!

**Comment by iangilman** [comment]
@JohnReagan Having the url be able to be a function seems like a great solution! That seems like it would be a good addition to the API. I'm not sure if there are any negative side effects... Perhaps places that don't expect the url to change? 

Would you be interested in putting together a PR to add something like this to the library?

**Comment by JohnReagan** [comment]
Hey @iangilman , thank you for your response! Yes, I'd be happy to get a PR together for this, although it will take some time to get bandwidth to do so.

**Comment by iangilman** [comment]
@JohnReagan Cool! No worries on the timing. And, of course, if someone else comes along who wants to tackle it first, they're welcome to do so!

**Comment by JohnReagan** [comment]
Hi @iangilman , I've put together a draft PR for review here when you have some time to review, thanks! https://github.com/openseadragon/openseadragon/pull/2247

**Comment by iangilman** [comment]
FIxed by #2247.
